### PR TITLE
Media query class for querying viewport dimensions

### PR DESCRIPTION
**Pull request by collinjackson**

### CODE REVIEW

File: sky/packages/sky/lib/src/material/material_app.dart [hunk 1]
```diff
@@ -73,18 +75,24 @@ class _MaterialAppState extends State<MaterialApp> {
...
-        child: new DefaultAssetBundle(
-          bundle: _defaultBundle,
-          child: new Title(
-            title: config.title,
-            child: new Navigator(
-              key: _navigator,
-              routes: config.routes,
-              onGenerateRoute: config.onGenerateRoute
+    return new SizeObserver(
+      onSizeChanged: (Size size) => setState(() { _size = size; }),
```
Comment (Hixie) [inline sky/packages/sky/lib/src/material/material_app.dart]: This won't work, it'll mean that the first frame, the MediaQuery will give bogus data.

Do we actually need the size of the app, or is the orientation of the device sufficient? The latter we can get via dart:ui, along with a notification system for handling changes.

File: sky/packages/sky/lib/src/widgets/media_query.dart [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+enum Orientation { portrait, landscape }
+
+class MediaQueryData {
+
+  const MediaQueryData({ this.size });
+
+  final Size size;
+
+  Orientation get orientation {
+    return screenSize.width > screenSize.height ? Orientation.landscape : Orientation.portrait;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: s/screenSize/size/

File: sky/packages/sky/lib/src/widgets/media_query.dart [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  String toString() => '$size $orientation';
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: I'd probably do something like '$runtimeType($size, $orientation)', so you can more easily tell what it is if you dump it without knowing what you're dumping (e.g. in a big Map or List).

File: sky/packages/sky/lib/src/widgets/media_query.dart [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'basic.dart';
+import 'framework.dart';
+import 'transitions.dart';
+
+enum Orientation { portrait, landscape }
+
+class MediaQueryData {
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: Needs an operator== and hashCode.

File: sky/packages/sky/lib/src/widgets/media_query.dart [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    Widget child
+  }) : super(key: key, child: child) {
+    assert(child != null);
+    assert(data != null);
+  }
+
+  final MediaQueryData data;
+
+  static MediaQueryData of(BuildContext context) {
+    Theme theme = context.inheritedWidgetOfType(Theme);
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: your copypasta is showing. :-)

**Comment by Hixie** [comment]
I think it's a useful building block for us to have.

File: sky/packages/sky/lib/src/material/material_app.dart (hunk 1 above)
Comment (collinjackson) [inline sky/packages/sky/lib/src/material/material_app.dart]: I think we actually need the size of the app, people are going to want to set breakpoints.

Comment (Hixie) [inline sky/packages/sky/lib/src/material/material_app.dart]: the size of the whole app, or the size of subsegments?

Comment (j9brown) [inline sky/packages/sky/lib/src/material/material_app.dart]: The "size of the app" isn't necessarily a well-defined concept.  The "app"
could be a full-screen activity or a widget in some larger activity.  I
think that policy over what layout to choose (eg. based on breakpoints and
other information) should be determined at a higher level and propagated
down through the app.  Maybe at the level of the router.

Jeff.

Comment (Hixie) [inline sky/packages/sky/lib/src/material/material_app.dart]: by "app" here I mean "Flutter Window". The area described by dart:ui.

**Comment by collinjackson** [comment]
Updated to add an example, PTAL

**Comment by collinjackson** [comment]
landscape
portrait

File: sky/packages/sky/lib/src/widgets/media_query.dart [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  final Size size;
+
+  Orientation get orientation {
+    return size.width > size.height ? Orientation.landscape : Orientation.portrait;
+  }
+
+  bool operator==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    MediaQueryData otherMediaQuery = other;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: call it typedOther, makes it easier to copy/paste later.

File: sky/packages/sky/lib/src/widgets/media_query.dart (hunk 4 above)
Comment (collinjackson) [inline sky/packages/sky/lib/src/widgets/media_query.dart]: Done

**Comment by Hixie** [comment]
LGTM.

**Pull request merged**